Penetration (and Gap) locator fixes

roystgnr opened the pull request:
Refs #31446, but this isn't a complete fix yet; I'm still hunting down a bug in large-deformation displaced-mesh cases.

## Reason
This is a fix for the issue described in #31446, at least for undeformed meshes and relatively small displacements.

Subsequent PRs will handle large deformations and hopefully make the whole fix more robust and automatic.

## Design

This adds a new `GeometricSearchInterface` parameter, `search_using_point_locator`, which allows users to switch to a search algorithm that doesn't depend on connectivity of the boundary being searched.

## Impact

Currently, this just adds a new user option for a more-robust-but-more-expensive search, without changing existing search behavior.

Eventually I'd like to add autodetection support for the option, but that will have to wait until I figure out how to autodetect the need robustly, and until I add and verify support for more sophisticated penetration options like normal-vector smoothing.

Thanks to @bwspenc for helping me understand the existing code.

roystgnr committed:
Flexible BreakMeshByElementGenerator bcid support

"Apply a boundary condition to only one side of a break or the other,
depending on the vagaries of element ids" doesn't really make sense to
me, but we'll leave it the default for backwards compatibility, while
adding the ability to add boundary conditions to both sides or neither
side of each break.

roystgnr committed:
Just use BoundaryInfo directly in PenetrationThread

This is quite a bit simpler and doesn't create temporary vectors.

It's theoretically possible that someone benchmarked the tuple
alternative, found a speedup that was worth the extra hassle, and just
forgot to comment that or report it upstream, but Occam says probably
not.

roystgnr committed:
Fix comment typos

roystgnr committed:
GeometricSearchInterface subs, use GSI validParams

Otherwise we can't add parameters there.

roystgnr committed:
Add option for penetration to use point_locator

Using the node_to_elem_map is fast, but in theory it could fail on 3D
AMR meshes (where a boundary node can be a hanging node unconnected to a
coarse element touching it), and in practice it fails on Flex IGA meshes
(where the boundary might be topologically disconnected at many nodes).

At this point I'm seeing some #31446 failures pass - 2D cases, for simple
things like PenetrationAux.

roystgnr committed:
Disconnected-mesh regression test for penetration

roystgnr committed:
Assertion in NearestNodeLocator output

This turned out to be a red herring for me, but the more assertions the
better.

roystgnr committed:
Don't use operator[] to read from a map

When there's missing entries, we don't need to construct key+nullptr
pairs to fill them.

roystgnr committed:
Allow PointLocator for quadrature penetration too

roystgnr committed:
Handling for ridge+peak cases with PointLocator

This doesn't yet do normal smoothing for penetration or calculate
tangential distances, but it should be enough to handle distance-only
situations like GapHeatTransfer.

roystgnr committed:
Recognize search_using_point_locator in ThContact

This gets the option working for me in GapHeatTransfer, and fixes at
least the earliest time steps of the cases whose errors prompted #31446

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/3191849/), step Docs: sync website on 339c4f3 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/31447/site)

This comment will be updated on new commits.

roystgnr commented:
Okay, *now* this PR should merely add a new option "without changing existing search behavior."  Assuming the basic CI agrees, though, I'll start appending more downstream recipes to be sure.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/3192733/), step Generate coverage on c24e88d wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>ae1258</th>
      <th colspan=4>#31447 c24e88</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.68%</td>
      <td>85.68%</td>
      <td>-0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>117827</td>
      <td>117908</td>
      <td>+81</td>
      <td>261</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19688</td>
      <td>19712</td>
      <td>+24</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31447/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31447/coverage/framework)


## Modules coverage

### Heat transfer

<table>
  <thead>
    <tr>
      <th></th>
      <th>ae1258</th>
      <th colspan=4>#31447 c24e88</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.19%</td>
      <td>89.20%</td>
      <td>+0.01%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>4611</td>
      <td>4615</td>
      <td>+4</td>
      <td>6</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>559</td>
      <td>559</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/31447/coverage_diff/heat_transfer)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/31447/coverage/heat_transfer)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/31447/coverage/xfem)

</details>



This comment will be updated on new commits.

moosebuild commented:
Job [Python 3.12](https://civet.inl.gov/job/3192764/) on c24e88d : invalidated by @roystgnr

Kicking Python 3.12 test first

roystgnr commented:
That python error isn't a fluke, huh?  Rebasing on devel.

loganharbour commented:
> That python error isn't a fluke, huh? Rebasing on devel.

FYI civet auto rebases on devel every time a job starts

loganharbour commented:
The python errors you're getting come from parsing the valid parameters in a `Diffusion` object (which now have an additional parameter from `GeometricSearchInterface` (which is `search_using_point_locator`).

You can regold the 19 to 20 in the python test that is failing (19 params -> 20, 1 new with your addition). The 3 to 4 is parameters that show up in a group. You added the `search_using_point_locator` to the default param group (so it shows up at the top). This probably isn't desired. Can you add your new param in `GeometricSearchInterface` to a new group, like:

```
InputParameters
GeometricSearchInterface::validParams()
{
  InputParameters params = emptyInputParameters();

  params.addParam<bool>(
      "search_using_point_locator",
      false,
      "Whether to use the mesh point locator (typically an octree search) to find "
      "elements around closest points of near-contact, contact, or penetration.  "
      "This is less efficient than the default search via node-element connectivity, but "
      "may be necessary to accurately detect gaps, contact, or penetration on any "
      "boundaries whose elements are not connected via shared nodes. ");
  params.addParamNamesToGroup("search_using_point_locator", "Geometric search");

  return params;
}
```

This should fix the 3 -> 4 diffs because the param is no longer a part of the default group.

roystgnr committed:
Add new parameter name to a group

roystgnr committed:
Regold python tests to match added parameter

loganharbour commented on modules/heat_transfer/src/actions/ThermalContactAction.C:
```diff
@@ -113,6 +113,15 @@ ThermalContactAction::validParams()
   params += GapConductance::actionParameters();
   params += GapConductanceConstant::actionParameters();
 
+  params.addParam<bool>(
```
Can you add this to a group too please

loganharbour reviewed (dismissed):
- I know the documentation is pretty bad, but can you add some guidance on the point locator flag in https://github.com/idaholab/moose/blob/next/framework/doc/content/source/geomsearch/PenetrationLocator.md?
- Take a look at the coverage diff. In `GeometricSearchData.C` the penetration locators never have the value set. Is this what you expect?
- The TRI case in `closest_point_to_side` isn't tested. Should it be?

roystgnr committed:
Add new ThermalContactAction parameter to a group

This hopefully cleans up the docs some

roystgnr committed:
Fix doc typo/atavism

roystgnr committed:
Fix atavism in GeometricSearchData doc

roystgnr committed:
Describe setUsePointLocator() in docs

roystgnr committed:
Doc GeometricSearchData setSearchUsingPointLocator

roystgnr committed:
Add more `search_using_point_locator` option docs

roystgnr committed:
Link GeometricSearchData docs where appropriate

roystgnr committed:
Add search_using_point_locator tet regression test

roystgnr committed:
BreakMeshByElementGenerator for "all subdomains"

This is easier and more robust than trying to be sure you remember every
id in an input file.

roystgnr committed:
Add "reverse" gap distance tests

I'm not sure which of these two behaviors is preferable, but I don't
want to change either of them without noticing and deliberately
regolding.

roystgnr committed:
Do p_info selection consistently w/point_locator

This gets us the gap to ridge points and peak points even when those
points are at the edge of a boundary rather than in between boundary
elements.  I think that's probably the right thing to do, but also it's
the easy thing to do.  ;-)

This requires regolding our first disconnected case, where we had a
primary boundary that was extending beyond its target.

GiudGiud commented on framework/src/meshgenerators/BreakMeshByElementGenerator.C:
```diff
@@ -25,17 +25,28 @@ BreakMeshByElementGenerator::validParams()
+      "interface_sides",
+      1,
+      "interface_sides<3",
+      "Whether to add no interface boundary, a 1-sided boundary, or a 2-sided boundary");
```
explain the rules for 1 sided? is it from smaller id to biger elem id?

roystgnr replied:
I will describe the rule (which you're correct about), but I have no explanation for why anyone would have wanted that behavior.  :wink:  I'm just keeping it as the default and making more sensible behaviors an option under the principle that breaking existing behavior without gradual deprecation is rude.

GiudGiud commented on framework/src/meshgenerators/BreakMeshByElementGenerator.C:
```diff
@@ -61,6 +76,10 @@ BreakMeshByElementGenerator::generate()
+  // We need to update the global_boundary_ids, and this is faster
+  // than a full prepare_for_use()
+  boundary_info.regenerate_id_sets();
+
```
I think to be consistent BMEG should be doing about the same thing as BBMG, e.g. be allergic to prepare_for_use to keep the previous connectivity. Which seems undesirable. 
It's fine as it is, just a thought

in fact looking at the code seems you are going  the other way

GiudGiud commented on framework/src/meshgenerators/BreakMeshByElementGenerator.C:
```diff
@@ -71,26 +90,28 @@ BreakMeshByElementGenerator::buildSubdomainRestrictedNodeToElemMap(
-    if (std::find(subdomains.begin(), subdomains.end(), elem->subdomain_id()) == subdomains.end())
+    // Skip if subdomains are specified and the element is not in them
+    if (!subdomains.empty() &&
+        std::find(subdomains.begin(), subdomains.end(), elem->subdomain_id()) == subdomains.end())
```
nice fix

GiudGiud commented on framework/src/meshgenerators/BreakMeshByElementGenerator.C:
```diff
@@ -151,11 +166,25 @@ BreakMeshByElementGenerator::createInterface(MeshBase & mesh,
+        boundary_info.add_side(elem_id, side, interface_id);
+  }
+
+  for (const auto & [elem_id, side] : sides_breaking)
```
```suggestion
  // Remove element neighbor connections on the broken sides
  for (const auto & [elem_id, side] : sides_breaking)
```

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -23,6 +23,74 @@
 using namespace libMesh;
 
+// Anonymous namespace for helper functions that ought to be moved
+// into libMesh
```
fairly sure we have this somewhere else in moose but also not sure where

GiudGiud replied:
might be in one of Liam's PR

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -23,6 +23,74 @@
+      const Point l01 = p1 - p0, l02 = p2 - p0;
+      const Point tri_normal = (l01.cross(l02)).unit();
+      const Point linecs = ((src - p0) * tri_normal) / tri_normal.norm_sq() * tri_normal;
+      Point in_plane = src - linecs;
+      Point planar_offset = in_plane - p0;
```
```suggestion
      const Point in_plane = src - linecs;
      const Point planar_offset = in_plane - p0;
```

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -190,248 +263,318 @@ PenetrationThread::operator()(const NodeIdRange & range)
+
+        p_info[best_i]->_closest_point = best_point;
+        p_info[best_i]->_distance =
+            (p_info[best_i]->_distance >= 0.0 ? 1.0 : -1.0) * std::sqrt(min_distance_sq);
```
negative distance?

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -190,248 +263,318 @@ PenetrationThread::operator()(const NodeIdRange & range)
               }
-              else // several ridges join at common node to make peak.  The common node is the
-                   // contact point
+              else // on a single ridge
```
all changes below are indentation right?

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -190,248 +263,318 @@ PenetrationThread::operator()(const NodeIdRange & range)
+        // backwards compatibility we're still using the older ridge
+        // and peak detection code, which depends on us ruling out
+        // sides that don't touch closest_node.
+        if (!_use_point_locator)
```
what is a ridge? an edge? A line between two nodes?

roystgnr replied:
In the existing code in 3D a "ridge" is the edge shared by two faces (which is also always an edge between two nodes; the existing code doesn't support non-conforming surface meshes) whose convex angle to each other means that a closest point can be on that line even though there's no orthogonal projection to either face; and a "peak" is the vertex shared by three or more faces whose convexity means that a closest point can be on that vertex even though there's no orthogonal projection to any of the edges connected to it.

I'd love for this to be better documented, but even with Ben's help I didn't trust myself to add much new information without inadvertently adding misinformation, and anyway I discovered that my new code path was going to be much simpler if it avoided the existing ridge and peak code entirely.

bwspenc replied:
The "ridge" and "peak" terminology is something I made up. @roystgnr described it correctly, but I'll say it in the way I was thinking of it: Think of a ridge on a roof or a mountain, with the faces being the two sides of the ridge, and then and being at a point above that ridge. If you're above the ridge, your projection points for the two faces are both off of those faces, but the contact point should be on the ridge, or the common edge in the FE terminology. There's a bunch of code that identifies the point on that ridge. Same basic idea for the peak. That code all applies for thermal contact, where the gap matters.

In contrast, if you're below the ridge, you project to both of the faces, and the point on the ridge is not needed. You instead have to determine which of the faces to enforce contact with, using the projection point on the face that has the strongest interaction. This is applicable for mechanical contact, which is enforced for penetrations.

I guess I could have just called it "edge" instead of "ridge", but the "ridge" terminology worked well for me to describe the scenario in easy-to-understand (to me at least) terms.

GiudGiud commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -190,248 +263,318 @@ PenetrationThread::operator()(const NodeIdRange & range)
-          {
-            bool foundSetWithMatchingNode = false;
-            for (unsigned int j = 0; j < ridgeSetDataVec.size(); ++j)
+          // Loop through all pairs of faces, and check for contact on ridge between each face pair
```
or is the ridge the contact line between two faces?

roystgnr committed:
Use clearer language than "incorporate"

roystgnr committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

roystgnr commented on modules/heat_transfer/src/actions/ThermalContactAction.C:
```diff
@@ -113,7 +113,17 @@ ThermalContactAction::validParams()
   params += GapConductance::actionParameters();
   params += GapConductanceConstant::actionParameters();
 
+  params.addParam<bool>(
```
Ignorance.  I'll give that a try.

roystgnr committed:
Avoid a one-parameter group

roystgnr committed:
Append rather than duplicating parameter code

roystgnr commented on modules/heat_transfer/src/actions/ThermalContactAction.C:
```diff
@@ -113,7 +113,17 @@ ThermalContactAction::validParams()
+
   params.addParamNamesToGroup("primary secondary", "Gap surface definition");
+  params.addParamNamesToGroup("search_using_point_locator", "Gap geometry search");
   params.addParamNamesToGroup(
```
It's using `applyParameters` in `addBCs`, which is where it's generating the `GapHeatTransfer` objects I was testing.  It looked like the tests were working there - is there anything else I need for those?

I'll add this to the lists for the aux kernels and dirac kernels.  Good save!

roystgnr committed:
Apply suggestions from code review

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

roystgnr committed:
Better document new BreakMeshByElement option

roystgnr committed:
Supply search_using_point_locator to more kernels

GiudGiud reviewed (commented):
more suggestions. chonky file more important than these

roystgnr committed:
Suggestions from Guillaume

Not applying these from the Github UI because I don't want to mess with
the heavy "rebase to get rid of huge gold files" changes I'm in the
middle of.

roystgnr commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -23,6 +23,74 @@
+    case QUAD4:
+    case QUAD8:
+    case QUAD9:
+      mooseError("Not implemented");
```
I'll add this manually to make my rebasing simpler

roystgnr commented on framework/src/geomsearch/PenetrationThread.C:
```diff
@@ -190,248 +263,318 @@ PenetrationThread::operator()(const NodeIdRange & range)
+          }
         }
+
+        p_info[best_i]->_closest_point = best_point;
+        p_info[best_i]->_distance =
```
It's fine.  That `switchInfo` call copies the `best_i` index into the `info` we care about, and the rest just gets deleted at the end.

Well, it's not *fine*, because we're still manually deleting raw pointers, but I'm trying to make this a minimal feature-addition PR; I could do a little code style cleanup separately later.

bwspenc commented on framework/src/geomsearch/GeometricSearchInterface.C:
```diff
@@ -20,7 +20,20 @@
+  InputParameters params = emptyInputParameters();
+
+  params.addParam<bool>(
+      "search_using_point_locator",
```
Yeah, that works for sure. I think you said that it's not always using octree, and if that's the case, maybe a more generic version could be "proximity"

bwspenc replied:
Hmm, I just remembered that the base algorithm also uses an octree for finding the set of nearest points, though. Maybe octree is ambiguous too.

roystgnr replied:
We've got non-octree options but in practice it's always an octree in 3D, quadtree in an XY plane mesh, binary in 1D.

But ... technically the K-D tree we always use for finding the initial nearest node isn't an octree, but it is pretty darn close to one, and either way I guess that "octree" is just a different *type* of internal jargon that users aren't necessarily going to recognize or be interested in.  We could switch `PointLocator` to use a K-D tree too and it wouldn't change anything about how this option works.

I'm liking "proximity" the more I think about it.  Really what's going on here is that we're getting the chance to find elements entirely via "geometry" rather than starting with geometry and expanding the search via "topology", but it would be weird to explain to users something like "this is the GeometricSearchInterface.  The search_using_geometry option is non-default".

But even the default algorithm still starts with proximity to find the closest node.  I'd be happy with "search_using_proximity", but what about "search_proximate_sides", since it's just the side search that's being changed?

roystgnr committed:
Regold test_appsyntax for parameter regrouping

roystgnr committed:
Clear references to geometric data doc

roystgnr committed:
Whitespace fix

GiudGiud reviewed (approved):
contingent on tests passing

roystgnr committed:
"search_using_point_locator" -> "search_method"

Ben recommended an enum here - "it communicates better that you are
choosing between two ways of doing it", "maybe an enumeration could be:
search_method = nearest_node_connected_sides | all_nearest_sides"

I liked that much better than any of the prior ideas either of us had
had, but I'm mixing it with his prior "search_all_proximate_sides"
suggestion because that better connotes "sides that are touching a
point" rather than just "sides that are nearest to the point".

roystgnr committed:
Correct docs to match new option name

bwspenc reviewed (approved):
My comments have been resolved.

roystgnr committed:
Use stable_sort in IndirectSort.h

std::sort doesn't give a uniquely defined result if it's given a partial
order that isn't a total order for its comparator.  SamplerBase doesn't
have any sort options that are total orders in general, and that was
causing our CSVDiffs to fail due to reordering in parallel and (at least
I'm hoping it was the same issue) on Mac.

std::stable_sort internally requests an extra memory buffer, so this
isn't free, but it should be worth it to prevent such surprises for
others in the future.

roystgnr commented on test/tests/geomsearch/3d_penetration_locator/gold/3d_disconnected_tet_reverse_out.e:
We're using SideValueSampler now, and going through the branch git log assures me that my rewrite of the branch history didn't leave any temporary copies of the 300kb files around.
